### PR TITLE
Use deep equality when comparing IngestDocument equality

### PR DESCRIPTION
**Pull request by dakrone**
This commit changes the `IngestDocument` class equality check to use "deep" equality. Previously we would end up doing the equivalent of `Objects.equals(thisMap, thatMap)` for the document. If a script or processor ended up writing any primitive arrays however, then those would fail the equality check because instance equality was used instead of check the actual values.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/ingest/IngestDocument.java [hunk 1]
```diff
@@ -908,7 +908,7 @@ public boolean equals(Object obj) {
         }
 
         IngestDocument other = (IngestDocument) obj;
-        return Objects.equals(ctxMap, other.ctxMap) && Objects.equals(ingestMetadata, other.ingestMetadata);
+        return Objects.equals(ctxMap, other.ctxMap) && Maps.deepEquals(ingestMetadata, other.ingestMetadata);
```
Comment (joegallo) [inline server/src/main/java/org/elasticsearch/ingest/IngestDocument.java:911]: I don't think `Maps.deepEquals(...)` happens to work recursively -- if you have string keys and long[] values, it'll do a good job, but if you have string keys and map values (where the maps are string keys to long[] values) then it doesn't work.

I think the consequence for us here is that nested field in an ingest document would still throw this off (e.g. `"somearray"` versus `"nested.somearray"`).

**Review by joegallo — CHANGES_REQUESTED** [review]
Github requires that I type words into this box (please see previously added comment-at-a-line).

File: server/src/main/java/org/elasticsearch/ingest/IngestDocument.java (hunk 1 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/ingest/IngestDocument.java:911]: Yes I agree. It doesn't work recursively. I will revisit this when I can.

**Comment by joegallo** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/ingest/IngestDocument.java (hunk 1 above)
Comment (joegallo) [reply server/src/main/java/org/elasticsearch/ingest/IngestDocument.java:911]: https://github.com/elastic/elasticsearch/pull/98901 has been merged which fixes `Maps.deepEquals` to handle this case correctly, and I've added a commit here to make sure we're exercising the nested maps codepath (which would have failed prior to #98901).

**Review by joegallo — APPROVED** [review]

**Pull request merged**